iceoryx_dds/include/iceoryx_dds/internal/gateway/dds_gateway_generic.inl

```diff
@@ -0,0 +1,160 @@
+{
+    auto channel = m_channelFactory(service);
+    m_channels->push_back(channel);
+    iox::dds::LogDebug() << "[DDSGatewayGeneric] Channel set up for service: "
+                         << "/" << service.getInstanceIDString() << "/" << service.getServiceIDString() << "/"
+                         << service.getEventIDString();
+    return channel;
+}
+
+template <typename channel_t, typename gateway_t>
+inline iox::cxx::optional<channel_t>
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::findChannel(const iox::capro::ServiceDescription& service) const
+    noexcept
+{
+    auto guardedVector = this->m_channels.GetScopeGuard();
+    auto channel = std::find_if(guardedVector->begin(), guardedVector->end(), [&service](const channel_t& channel) {
+        return channel.getService() == service;
+    });
+    if (channel == guardedVector->end())
+    {
+        return iox::cxx::nullopt_t();
+    }
+    else
+    {
+        // Needs to be dereferenced since the smart lock returns pointers to elements in concurrent collections.
+        return iox::cxx::make_optional<channel_t>(*channel);
+    }
+}
+
+template <typename channel_t, typename gateway_t>
+inline void
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::forEachChannel(const std::function<void(channel_t&)> f) const
+    noexcept
+{
+    auto guardedVector = m_channels.GetScopeGuard();
+    for (auto channel = guardedVector->begin(); channel != guardedVector->end(); ++channel)
+    {
+        f(*channel);
+    }
+}
+
+template <typename channel_t, typename gateway_t>
+inline void iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::discardChannel(
+    const iox::capro::ServiceDescription& service) noexcept
+{
+    auto guardedVector = this->m_channels.GetScopeGuard();
+    auto channel = std::find_if(guardedVector->begin(), guardedVector->end(), [&service](const channel_t& channel) {
+        return channel.getService() == service;
+    });
+    if (channel != guardedVector->end())
+    {
+        guardedVector->erase(channel);
+        iox::dds::LogDebug() << "[DDSGatewayGeneric] Channel taken down for service: "
+                             << "/" << service.getInstanceIDString() << "/" << service.getServiceIDString() << "/"
+                             << service.getEventIDString();
+    }
+}
+
+// ================================================== Private ================================================== //
+
+template <typename channel_t, typename gateway_t>
+inline void iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::discoveryLoop() noexcept
+{
+    iox::dds::LogDebug() << "[DDSGatewayGeneric] Starting discovery.";
+    while (m_isRunning.load(std::memory_order_relaxed))
+    {
+        iox::capro::CaproMessage msg;
+        while (this->getCaProMessage(msg))
+        {
+            discover(msg);
+        }
+        std::this_thread::sleep_until(std::chrono::steady_clock::now()
```

Comment:
why not just `sleep_for`?

Reply:
sleep_until is more precise if we take the time with now() exactly after we wake up and then add the desired waiting period t, execute what we intend to (which should be doable in t, otherwise we are already in delay), and then wait until the calculated time